nexus/pkg/filestream/loop_transmit.go

```diff
@@ -0,0 +1,161 @@
+package filestream
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/go-retryablehttp"
+)
+
+var completeTrue bool = true
+
+type chunkData struct {
+	fileName string
+	fileData *chunkLine
+	Exitcode *int32
+	Complete *bool
+}
+
+type chunkLine struct {
+	chunkType ChunkFile
+	line      string
+}
+
+// FsChunkData is the data for a chunk of a file
+type FsChunkData struct {
+	Offset  int      `json:"offset"`
+	Content []string `json:"content"`
+}
+
+type FsData struct {
+	Files    map[string]FsChunkData `json:"files,omitempty"`
+	Complete *bool                  `json:"complete,omitempty"`
+	Exitcode *int32                 `json:"exitcode,omitempty"`
+}
+
+func (fs *FileStream) addTransmit(chunk chunkData) {
+	fs.transmitChan <- chunk
+}
+
+func (fs *FileStream) loopTransmit(inChan <-chan chunkData) {
+
+	overflow := false
+
+	for active := true; active; {
+		var chunkMaps = make(map[string][]chunkData)
+		select {
+		case chunk, ok := <-inChan:
+			if !ok {
+				active = false
+				break
+			}
+
+			chunkMaps[chunk.fileName] = append(chunkMaps[chunk.fileName], chunk)
+
+			delayTime := fs.delayProcess
+			if overflow {
+				delayTime = 0
+			}
+			delayChan := time.After(delayTime)
+			overflow = false
+
+			for ready := true; ready; {
+				select {
+				case chunk, ok = <-inChan:
+					if !ok {
+						ready = false
+						active = false
+						break
+					}
+					chunkMaps[chunk.fileName] = append(chunkMaps[chunk.fileName], chunk)
+					if len(chunkMaps[chunk.fileName]) >= fs.maxItemsPerPush {
+						ready = false
+						overflow = true
+					}
+				case <-delayChan:
+					ready = false
+				}
+			}
+			for _, chunkList := range chunkMaps {
+				fs.sendChunkList(chunkList)
+			}
+		case <-time.After(fs.heartbeatTime):
+			for _, chunkList := range chunkMaps {
+				if len(chunkList) > 0 {
+					fs.sendChunkList(chunkList)
+				}
+			}
+		}
+		if fs.stageExitChunk != nil {
+			fs.sendChunkList([]chunkData{*fs.stageExitChunk})
+		}
+	}
+}
+
+func (fs *FileStream) sendChunkList(chunks []chunkData) {
+	var lines []string
+	var complete *bool
+	var exitcode *int32
```

Comment:
can we use a new chunk instead it already has all these fields.

Reply:
yes, this was a reorg, but not changing the code too much, the next change is redoing sendChunkList